DOC: Added new page describing how to contribute to the docs team

melissawm opened the pull request:
This page is a guide to writing documentation for numpy. A guide to writing tutorials (a missing link in this document) is coming in a separate PR.

Closes gh-14817

ChrisBarker-NOAA committed:
DOC: a few typos and copy editing in NEP 30

Co-Authored-By: Peter Andreas Entschev <peter@entschev.com>

pentschev committed:
NEP: Proposal for array creation dispatching with `__array_function__`

This NEP proposes the introduction of a `like=` argument. The argument is to be
added to all array creation functions. This allows the user to pass a reference
array that will use `__array_function__` protocol to dispatch such functions to
downstream libraries.

pentschev committed:
MAINT: fix implementation example of NEP-33

pentschev committed:
MAINT: cite few examples of array creation functions in NEP-33

pentschev committed:
MAINT: change array creation dispatching with `__array_function__` to NEP-35

Commit:
DOC: document inconsistent mask results in MaskedArray

Co-Authored-By: Tianxiang “Ronnie” Gao <tienhsiangkao@tamu.edu>

Balandat committed:
Bug: Fix eigh mnd cholesky methods of numpy.random.multivariate_normal

Fixes #15871

Commit:
BUG: add check if pvals is 1d array in numpy.random.*.multinomial

Commit:
Add testing and replace check

Commit:
Replace checks with direct function call

anirudh2290 committed:
DOC: Correct private function name to PyArray_AdaptFlexibleDType

mattip committed:
BLD: report clang version on macOS

mattip committed:
BLD: use xcode10 for one CI run

melissawm committed:
Added guide to documentation contributions.

melissawm committed:
Added guide to documentation contributions.

melissawm committed:
Added toctree entry for new document.

melissawm committed:
Added toctree entry for new document.

mattip committed:
Update doc/source/reference/maskedarray.baseclass.rst

Co-Authored-By: Sebastian Berg <sebastian@sipsolutions.net>

mattip commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+
+- **Tutorials, How Tos, Explanations:** These documents are stored in the NumPy
+  source code tree, which means that in order to add them to the official
+  documentation, you have to download the NumPy source code, build it and
```
maybe needs a link to [``howto-build-docs``](https://numpy.org/devdocs/docs/howto_build_docs.html)

mattip committed:
Merge pull request #15749 from panpiort8/masked_array_docs

DOC: document inconsistency between the shape of data and mask in numpy.ma

charris committed:
Merge pull request #15872 from Balandat/fix_eigh_mvn_sampling

BUG: Fix eigh and cholesky methods of numpy.random.multivariate_normal

rgommers committed:
Merge pull request #15903 from mattip/report-clang

BLD: report clang version on macOS

melissawm commented:
Hi all, I've just found [this issue](https://github.com/numpy/numpy/issues/14817) and I think there are some valid points there that I didn't include in this PR, so I'll be revising it shortly. I'll let you know once I'm finished. Thanks!

WarrenWeckesser committed:
BUG: random: Disallow p=0 in negative_binomial

Make `Generator.negative_binomial` raise a ValueError if p=0.

`negative_binomial(n, p)` draws samples from the distribution of
the number of failures until n successes are encountered. If p is 0,
then a success is never encountered, so the probability distribution
is 0 for any finite number of failures. In other words, it is not
really a meaningful distribution, so we disallow p=0.

Closes gh-15913.

Commit:
Add minor tests enhancement

rgommers commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+<https://mail.python.org/mailman/listinfo/numpy-discussion>`_. Meeting minutes
+are stored `on hackmd.io <https://hackmd.io/oB_boakvRqKR-_2jRV-Qjg>`_.
+
+You can find a summary of our long term goals for the documentation `at our
```
How about "larger planned and in-progress documentation improvment ideas"? It's not really a summary of long-term goals (the below text in this PR is better for that).

rgommers commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+
+**Where is the documentation?**
+
+The main page for the :ref:`NumPy Documentation <manual>` lists several
```
unfinished link? if it does exist somewhere, could you change it to be longer so it's easier to locate via (e.g.) `grep`?

melissawm replied:
This is generated from the `contents.rst` file, through a template. However, I see that when it's rendered with that reference it's not the same as it shows up on https://numpy.org/devdocs`. I'm not really sure how to link it locally, except linking it to `doc/build/html/index.html`. What would be the best approach here?

rgommers replied:
How about simply renaming the label from `_manual` to `_numpy_docs_mainpage`?

There's only one other place where the label is used that should also be renamed, in `source/reference/index.rst`. And if one is missed you'll see `WARNING: undefined label` and CI will fail.

The main page title is now `NumPy manual contents`, and "manual" is a bit of a weird name nowadays. So second suggestion: rename that to simply `NumPy documentation`.

melissawm replied:
Agreed, the "Manual" was bothering me too. Thanks!

rgommers commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+If you have already decided which type of document you want to write, you can
+check out the following specific guides:
+
+- Guide to writing Tutorials
```
maybe add a `(TODO)` here for now

rgommers commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+Other ways to contribute
+------------------------
+
+If you see a good tutorial, how to or explanation that is not included in the
```
"how to" -> "how-to" if it's a verb? reads less confusing to me

rgommers replied:
same in a few more places in this document

rgommers commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+
+If you see a good tutorial, how to or explanation that is not included in the
+official documentation, you can suggest it to be added by `opening an issue on
+github <htpps://github.com/numpy/numpy/issues>`_. Similarly, opening issues to
```
github -> GitHub

rgommers replied:
and same further down

anirudh2290 replied:
htpps://github.com/numpy/numpy/issues -> https://github.com/numpy/numpy/issues

rgommers commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+
+**Final tips**
+
+- Don't worry if english is not your first language. Do your best - we'll revise
```
english - English

rgommers reviewed (commented):
Thanks @melissawm, this is great! Only minor comments, the structure here is nice. Incorporating some of the comments on gh-14817 also sounds good, those all make sense to me.

charris commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+or just need someone to guide you through your first steps - we're always happy
+to help. Meetings are usually announced on the `numpy-discussion mailing list
+<https://mail.python.org/mailman/listinfo/numpy-discussion>`_. Meeting minutes
+are stored `on hackmd.io <https://hackmd.io/oB_boakvRqKR-_2jRV-Qjg>`_.
```
Needs two trailing underscores, also above. Might be worth running a search for `_ to check for those.

anirudh2290 commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+  submitting an issue on github suggesting it, or asking on the mailing
+  list/slack channel.
+- If you are unfamiliar with git/github or the process of submitting a pull
+  request (PR), check our :ref:`Contributor Guide <devindex>`. 
```
nit: remove whitespace at the end.

anirudh2290 commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+developers and contributors, as well as discussions, presentations, videos and
+other user generated content. It may include learning-oriented content (such as
+Tutorials and How Tos), Use Cases or in depth explanations and reference for
+developers. 
```
nit: remove whitespace at the end

anirudh2290 commented on doc/source/dev/howto-docs.rst:
```diff
@@ -0,0 +1,109 @@
+(that is, the documentation that ships with NumPy and lives on the official
+project pages). Keep in mind that if you don't want to do this, writing a
+tutorial on your own blog, creating a youtube video or answering questions on
+social media or the slack channel are also great contributions! 
```
nit: remove whitespace at the end of the line.

mattip committed:
ENH: use sysconfig instead of probing Makefile

mattip committed:
Merge pull request #15914 from WarrenWeckesser/negbin-pgt0

BUG: random: Disallow p=0 in negative_binomial

mattip committed:
ENH: update default MACOSX_DEPLOYMENT_TARGET to 10.9 (from review)

mattip committed:
Merge pull request #15931 from anirudh2290/nep_doc_correction

DOC: Correct private function name to PyArray_AdaptFlexibleDType

feeeper committed:
Typo in the `mtrand.pyx` was fixed

`ValueError` about negative dimension was lowercased.

charris committed:
Merge pull request #15936 from feeeper/typo-in-value-error-in-random-choice

MAINT: Fix capitalization in error message in `mtrand.pyx`

m-podhajski committed:
TST: Parametrize sort test (#15806)

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>
Co-authored-by: Matti Picus <matti.picus@gmail.com>

Closes #13528

hameerabbasi committed:
BUG,DEP: Make `scalar.__round__()` behave like pythons round (#15840)

See issue gh-15297 and related mailing list discussion.

This PR bring scalar.__round__() in line with python, so that `round(scalar)` always returns a python integer, while `round(scalar, ndigits=0)` returns the same type. Since complex numbers are not supported in Python, and cannot be reasonably cast to integers they are deprecated. 

Closes gh-15297

seberg committed:
DOC: Update documentation w.r.t. NPY_RELAXED_STRIDES_CHECKING (gh-15907)

This has been the default for a very long time, so this was
incorrect and mentioning any versions is probably not helpful.

pdebuyl committed:
DOC: Improve docstring for numpy.linalg.lstsq (gh-15920)

Start the docstring with """Computes the vector x that approximatively
solves the equation :math:`a x = b`.""" instead of """Solves the
equation :math:`a x = b`""".

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

Closes gh-15874

WarrenWeckesser committed:
Merge pull request #15921 from mattip/use-sysconfig

ENH: Use sysconfig instead of probing Makefile

anirudh2290 committed:
TST: Add slow_pypy support

vrakesh committed:
DOC:Fix method documentation of function sort()

mattip committed:
Merge pull request #15876 from panpiort8/multinomial_pvals_1d

BUG: Check that `pvals` is 1D in `_generator.multinomial`.

rgommers committed:
Merge pull request #14715 from pentschev/nep-like-parameter

NEP: Proposal for array creation dispatching with `__array_function__`

rgommers committed:
DOC: rename NEP 34 file for consistency

rgommers committed:
DOC: fix a NEP reference, number was incorrect (36 -> 35); update link

mattip committed:
Merge pull request #15953 from rgommers/neps-maint

NEP: minor maintenance, update filename and fix a cross-reference

rgommers committed:
Merge pull request #14530 from ChrisBarker-NOAA/ChrisBarker-NEP-30-copy-editing

MAINT:  Fix typos and copy edit NEP-0030.

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.8.0 to 5.8.3

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.8.0 to 5.8.3.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.8.0...hypothesis-python-5.8.3)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

seberg committed:
Merge pull request #15964 from numpy/dependabot/pip/hypothesis-5.8.3

MAINT: Bump hypothesis from 5.8.0 to 5.8.3

melissawm committed:
Added note to angle function docstring about angle(0) being 0.

melissawm committed:
Added comments from gh-14817. Added suggestions from PR review.

melissawm committed:
Added comments from gh-14817. Added suggestions from PR review.

melissawm commented:
Thanks for the comments! Here's the new version. Let me know if you have any other suggestions.

mattip committed:
Merge pull request #15968 from melissawm/angle-docs

DOC: Added note to angle function docstring about angle(0) being 0.

eric-wieser commented on doc/source/dev/howto-docs.rst:
```suggestion
about documentation, `NEP 44
<https://www.numpy.org/neps/nep-0044-restructuring-numpy-docs>`__.
```

anirudh2290 committed:
MAINT: Make if else clause more readable

melissawm committed:
Changed link/label to main docs page

melissawm committed:
Changed link/label to main docs page

anirudh2290 committed:
MAINT: Remove 0d check for PyArray_ISONESEGMENT

eric-wieser committed:
MAINT: Address a TODO that was waiting for the umath/multiarray merge

eric-wieser committed:
MAINT: Move declarations next to usage

eric-wieser committed:
BUG: Fix double Py_DECREF

`PyArray_SetWritebackIfCopyBase` steals a reference to base, even if it fails.

eric-wieser committed:
BUG: Avoid duplication in stack trace of `linspace(a, b, num=1.5)`

Before this commit, the stack trace was:
```
Traceback (most recent call last):
  File "C:\Users\wiese\Repos\numeric-python\numpy\build\testenv\Lib\site-packages\numpy\core\function_base.py", line 114, in linspace
    num = operator.index(num)
TypeError: 'float' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "<ipython-input-13-802b8c6e85f6>", line 1, in <module>
    np.linspace(1, 2, 1.5)
  File "<__array_function__ internals>", line 5, in linspace
  File "C:\Users\wiese\Repos\numeric-python\numpy\build\testenv\Lib\site-packages\numpy\core\function_base.py", line 116, in linspace
    raise TypeError(
TypeError: object of type <class 'float'> cannot be safely interpreted as an integer.
```

Now it is
```
Traceback (most recent call last):
  File "C:\Users\wiese\Repos\numeric-python\numpy\build\testenv\Lib\site-packages\numpy\core\function_base.py", line 114, in linspace
    num = operator.index(num)
TypeError: 'float' object cannot be interpreted as an integer
```

This noisy traceback was introduced in f4dfe833e3e037bb69113f7250fad3699f918cfc.

seberg committed:
Merge pull request #15985 from eric-wieser/shorten-operator.index-backtrace

BUG: Avoid duplication in stack trace of `linspace(a, b, num=1.5)`

mattip committed:
Merge pull request #15967 from anirudh2290/slow_pypy_impl

TST: Add slow_pypy support

seberg committed:
Merge pull request #15982 from eric-wieser/address-todo

MAINT/BUG: Cleanup and minor fixes to conform_reduce_result

anirudh2290 committed:
NEP: Improve Value Based Casting paragraph in NEP 40 (#16000)

* DOC: Improve Value Based Casting doc

* Update doc/neps/nep-0040-legacy-datatype-impl.rst

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

* Update doc/neps/nep-0040-legacy-datatype-impl.rst

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

seberg committed:
Merge pull request #15989 from anirudh2290/remove_0d_check

MAINT: Remove 0d check for PyArray_ISONESEGMENT

dustanlevenstein committed:
DOC: Improve record module documentation (#15899)

Addresses parts of #15853 by adding full function documentions with parameters to
`numpy/core/records.py`.

Co-authored-by: Dustan Levenstein <dlevenstein@gmail.com>
Co-authored-by: Eric Wieser <wieser.eric@gmail.com>
Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

seberg committed:
Merge pull request #15993 from vrakesh/masked_array_sort_doc

DOC: Fix method documentation of function sort in MaskedArray

pdebuyl committed:
DOC: add note on flatten ordering in matlab page (#16001)

Add note on the ordering of matlab's ``x(:)`` vs NumPy ``x.flatten()``

fix #15837

adam4130 committed:
BUG: Fix inf and NaN-warnings in half float `nextafter` (#15988)

This commit fixes the at `nextafter(np.inf, ...)` should not return NaN unless the second value is NaN. Also `nextafter(NaN, ...)` and `nextafter(..., NaN) should not give the invalid value warning (no invalid value is created, the invalid value is propagated).

This aligns half-float `nextafter` with double/float `nextafter`.

Closes #15977

panpiort8 committed:
BUG: Alpha parameter must be 1D in `generator.dirichlet` (#15951)

Only one dimensional alpha paramter is currently supported, but higher dimensions were silently allowed and gave an incorrect results. This fixes the regression. In the future, the API could be extended to allow higher dimensional arrays for alpha.

Fixes gh-15915

charris committed:
REL: Update master after 1.18.3 release.

rgommers committed:
Merge pull request #16021 from charris/update-after-1,18.3-release

REL: Update master after 1.18.3 release.

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.8.3 to 5.10.1

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.8.3 to 5.10.1.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.8.3...hypothesis-python-5.10.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

seberg committed:
Merge pull request #16024 from numpy/dependabot/pip/hypothesis-5.10.1

MAINT: Bump hypothesis from 5.8.3 to 5.10.1

rossbar commented on doc/source/dev/howto-docs.rst:
IMO the wording here is a little confusing: it seems to imply that the *reference documentation* is built when NumPy itself is built from source, and doesn't mention the extra docs-building step of converting the docstrings in the source code to the rendered HTML (i.e. building *the docs* from source).

I don't think this is a major issue, but it could be confusing to those who are not familiar with the distinction between building NumPy and building the docs.

rossbar commented on doc/source/dev/howto-docs.rst:
```suggestion
documentation within the project. You can read the
```

rossbar commented on doc/source/dev/howto-docs.rst:
The link title (i.e. "Sphinx Quick Reference Guide") is a little misleading as it actually points to the rST quick reference. Also, this link certainly has all the information, but the docutils docs are so ugly... maybe consider linking to sphinx's rST primer (http://www.sphinx-doc.org/en/stable/usage/restructuredtext/basics.html) instead?

melissawm replied:
I'll do both, if that's ok. I think it's better to have more sources of information when possible.

rossbar commented on doc/source/dev/howto-docs.rst:
```suggestion
- :ref:`Guide to writing Reference (API) documentation: the numpydoc 
   docstring guide <howto-document>`
```
Also, should "Reference" be capitalized, given the pattern used in other titles?

rossbar commented on doc/source/dev/howto-docs.rst:
Hopefully my review comments aren't serving as *too much* evidence for this point :grimacing:

melissawm replied:
Not at all, thanks for those, they were very helpful :)

rossbar reviewed (commented):
This looks really great, thanks @melissawm.

I had a couple comments of the type-editing variety. The content itself is very informative and welcoming

paulrougieux committed:
DOC: Clarify docs on mixed advanced indexing and slicing (gh-15891)

It was not clear how to reproduce the slice example with multi dimensional array indexing. I hope the additional example clarifies the equivalence between slice and multi array for new users.

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>
Co-Authored-By: Anirudh Subramanian <anirudh2290@apache.org>

rgommers committed:
DOC: update first section of NEP 37 (``__array_function__`` downsides) (#16015)

* DOC: update first section of NEP 37 (``__array_function__`` downsides)

[ci skip]

* Update NEP 37 based on review comments on gh-16015

melissawm commented on doc/source/dev/howto-docs.rst:
As far as I know that page does not yet exist, but it will on the new website. So I'll do that once the website is up. Is that ok?

rossbar replied:
Of course! I couldn't find a link either otherwise I would've explicitly suggested it

melissawm committed:
Changes from PR discussions.

melissawm committed:
Merge branch 'howtodocs' of github.com:melissawm/numpy into howtodocs

charris commented:
Looks like you merged from master and picked up a bunch of unrelated commits. They will probably go away when merged, but it is usually better to rebase than merge.

charris commented:
Note that you now have 74 commits affecting 44 files.

melissawm commented:
Oh wow. Don't know what happened there, I'm sorry. I actually rebased because there was some conflict but I'm not sure what went wrong. Should I undo this somehow, or do a new PR? How do I fix this? Again, sorry!

charris commented:
@melissawm Those extra commits will disappear once merged to master, the main effect is to make review difficult. The easiest thing might be to open a new PR off current master and merge this PR branch to it. Seems to work.
```
charris@fc [numpy.git (pr/15908)]$ git co -b test master
Switched to a new branch 'test'
charris@fc [numpy.git (test)]$ git merge pr/15908
Updating 8dfaa4a5e0..cf9a993cc6
Fast-forward
 doc/source/contents.rst        |   8 +--
 doc/source/dev/howto-docs.rst  | 145 +++++++++++++++++++++++++++++++++++++++++
 doc/source/dev/index.rst       |   3 +
 doc/source/reference/index.rst |   2 +-
 4 files changed, 153 insertions(+), 5 deletions(-)
 create mode 100644 doc/source/dev/howto-docs.rst
```

melissawm commented:
@charris Thanks for the help. That works, but now should I close this PR and open a new one, or is there a way to fix this one here? I appreciate your help and patience here :)

charris commented:
Close this and open a new one. The comments will be lost, but this PR seems pretty much done anyway. Just add a reference to this in the new PR.

melissawm commented:
Thanks again. I'm closing this and heres the new one #16069
